File: lintcode/Minimum_Path_Sum.cpp
class Solution {
public:
    /**
     * @param grid: a list of lists of integers.
     * @return: An integer, minimizes the sum of all numbers along its path
     */
    int minPathSum(vector<vector<int> > &grid) {
        if(grid.size() == 0) throw runtime_error("min path does not exist!");
        
        int m = grid.size();
        int n = grid[0].size();
        
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(i == 0 && j > 0) grid[i][j] += grid[i][j-1];
                if(j == 0 && i > 0) grid[i][j] += grid[i-1][j];
                if(i > 0 && j > 0) grid[i][j] = min(grid[i-1][j], grid[i][j-1]) + grid[i][j];
            }
        }
        return grid[m-1][n-1];
    }
};

File: lintcode/java/MinimumPathSum.java
public class Solution {
    /**
     * @param grid: a list of lists of integers.
     * @return: An integer, minimizes the sum of all numbers along its path
     */
    public int minPathSum(int[][] grid) {
        if(grid.length == 0) return 0;
        
        int m = grid.length;
        int n = grid[0].length;
        
        int[][] t = new int[m][n];
        
        for(int i = 0; i < m; i++){
            for(int j = 0; j < n; j++){
                if(i == 0 && j == 0) t[i][j] = grid[i][j];
                else if(i == 0 && j > 0) t[i][j] = t[i][j-1] + grid[i][j];
                else if(i > 0 && j == 0) t[i][j] = t[i-1][j] + grid[i][j];
                else t[i][j] = Math.min(t[i-1][j], t[i][j-1]) + grid[i][j];
            }
        }
        
        return t[m-1][n-1];
    }
}

